### DISCUSSION TITLE
switch expressions with modulo are reported as not exhaustive when, in fact, they are.

### DISCUSSION THREAD

**Discussion by mphipps1** [dotnet/csharplang#5678] [Language Ideas]
**Version Used**: 
VS2022 Preview / C#10
**Steps to Reproduce**:
Create a C#10 program with this code:
void test()
{
    var i = (uint) new Random().Next();
    var j = (i % 4) switch
    {
        0 => 0,
        1 => 1,
        2 => 1,
        3 => 1,
    };
}

**Expected Behavior**:
Switch is OK because all cases covered.

**Actual Behavior**:
C58509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern "4U' is not covered.
The 'switch' expression does not handle all possible inputs (it is not exhaustive). For example, the pattern '4U' is not covered.

**Comment by mphipps1** [comment]
I am unsure why this was moved from an issue. The compiler is warning about something that it should not be.

**Reply by HaloFour** [reply]
Issues are reserved for proposals with full specifications that the team is moving through the design process.  At this point this is a request for a language change.

In this case the request is that the compiler evaluate the code to come up with a possible limit of values for the purpose of exhaustiveness checking, which the compiler does not so.

**Reply by 333fred** [reply]
It was moved because this is a request for a new language feature, not an issue with roslyn. The compiler is behaving exactly as specified: the input of the switch expression is an unsiged integer, so there are (2^32)-4 possible values that are not being handled by that switch expression. Yes, we as humans can look at that and determine that, for this simple case, there can only be 0-3, but this is not easy for a compiler to do. Consider a simple refactoring:

```cs
void test()
{
var i = (uint) new Random().Next();
var j = GetRemainder(i) switch
{
0 => 0,
1 => 1,
2 => 1,
3 => 1,
};
}

void GetRemainder(int i) => i % 4;
```

Now the compiler would need to look at the body of `GetRemainder` to determine if there was any info to narrow the range. This gets exponentially more complex, and solving it generally is effectively trying to solve the halting problem. Because the compiler can't do that, we don't make simple cases like this work either.

**Reply by CyrusNajmabadi** [reply]
I apologize.  I moved this, and i'd written up a message as to why.  But then a meeting started and i forgot to publish the message.  Others have mentioned correctly why this was moved, so i won't repeat taht :)

**Comment by mphipps1** [comment]
No worries  Cyrus. Thanks for responding.  Your kindness and humility are humbling, honestly.

I don't know what the right answer is, here. 
I don't like the warning. I always consider it a code smell to have warnings in my code. I take them seriously and am infinitely grateful to you guys for making them!
So, the alternative is a default case. That feels weird, honestly. It clutters up the code. I have taken to making them, none the less. 
This makes me think about Pascal's ranged integers. That's some old school tech that managed this kind of thing. On the other hand, they were a PAIN IN THE NECK, sometimes. As were Pascal's arrays as strings, for similar reasons.
I would be happy (happier) with a solution that attempts to validate the range and then warns if not.

**Reply by FaustVX** [reply]
I don't think warning are code smell, it's just the compiler telling you "Are you sure this is the right behavior for this piece of code ?"
And in your case, you are okay with the fact there are only 4 values to match.
You could ignore the warning with `#pragma warning disable warning CS8509 //The switch expression does not handle all possible values of its input type (it is not exhaustive)`

Then a `#pragma warning restore CS8509` after the `switch`.
[SharpLab.io](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABATARgFgAobAZgAI8KBhCgbxIuasuwBYKBZACgEoGJAMQAHKAEMA5gFtxFAO7ioAOxjLJFFDADO4gEYAbCLQDKADgCsABgCcAejsUAKgAtj2+TGABjFxQgAHmIQ2towAPbKmuEhFMrhwBQu4sooRhTiBgYUIuGhMIbGYJkArrHhcBRe2lXKIiWJwAgixjxeVTXxiYHJJdrAMJB8TCyjxVBVFAC8FDwWAKTsAh5eviOjzIzEGzsbVlMAfADkVkdI67s7+IdH+GcXl6O4N7j32487ZDdkb5cAvgBuYRiKSyBRKVTqChQELAcIw0yWWwXP4kP5AA=)

**Reply by FaustVX** [reply]
And also, with these discussions: #1070 #5648, you could do `#pragma warning disable once CS8509 //The switch expression does not handle all possible values of its input type (it is not exhaustive)` before the `switch`

**Reply by sab39** [reply]
I think the idea is that you'd still want to get a warning if you switch on `i % 4` but only handle cases 0, 1 and 2. It would be lovely but I imagine pretty difficult to specify reliably!

**Reply by FaustVX** [reply]
I'm okay with your idea, but the use case is so small.
Overall, I don't think it's that useful.

**Reply by DavidArno** [reply]
@mphipps1,

Rather than having a separate default case, you can combine that with the last case using an or pattern:
```c#
var j = (i % 4) switch {
    0 => 0,
    1 => 1,
    2 => 1,
    3 or _ => 1,
}; 
```

**Comment by CyrusNajmabadi** [comment]
I'm not opposed to the language getting smarter about integral subset regions.  @RikkiGibson for thoughts here as well.  That said, i don't think it's majorly important and the status quo is fine.

**Reply by 333fred** [reply]
Personally, I'd only want to see this if we added some kind of linear type system, where `integer % integer` returned some kind of `int constraint to range` type that could then flow through the rest of the type system. But... that's a lot of work.

**Reply by RikkiGibson** [reply]
If users wish to do this on a "best effort" basis, they could write a diagnostic suppressor.

**Reply by RikkiGibson** [reply]
I could also squint and imagine some kind of "active patterns" feature addressing this.

Maybe a "modulo pattern" of some kind (*extreme* strawman):

```cs
int M(int x)
{
    return x switch
    {
        % 4 == 0 => ...
        % 4 == 1 => ...
        % 4 == 2 => ...
        % 4 == 3 => ...
    }; // no warning: we can tell it's exhaustive
}
```

I have actually wanted something similar before to check flags enums, but I couldn't come up with a syntax that felt even worth considering 🙁:

```cs
bool M(MyEnum x)
{
    return x switch
    {
        // similar to checking `(x & MyEnum.Flag1) != 0`
        (& MyEnum.Flag1) != 0 => ...
    }
}
```

**Reply by mphipps1** [reply]
The thing that I always appreciated about the ranged integer approach is that it lets you disable/ignore some checks. Array bounds, divide by 0, overflow and underflow, for example. You can elide the checks if the integer is provably not going to cause a problem. That is a fair amount of work, though, tagging each integer node in the AST with ranges. And, of course, you can't always tell (easy example:: Integer.Parse(someStringIReadFromConsole); )
But, when you can, you can enhance performance and eliminate classes of bugs. 
I don't have any idea how often something like this would be useful and applicable, though, or how it would impact compile performance.

**Reply by sab39** [reply]
I think - at least for this particular case - I'd like it *less* with a full fledged ranged integer type, because it works so hard to solve the problem but still falls short all over the place. For example, `switch ((x % 4) * 2)` could certainly be smart enough to know that the range is 0..8, but it's still going to force you to handle 1, 3, 5 and 7. Having some specific simple scenarios like `expr % const` get special cased seems easier to understand for developers than an elaborate, complex set of rules that are hard to reason about. One way or another you *will* sometimes have to handle nonexistent cases to keep the compiler happy, so it's better to have clear and simple criteria for when.

**Reply by CyrusNajmabadi** [reply]
> Personally, I'd only want to see this if we added some kind of linear type system, where integer % integer returned some kind of int constraint to range type that could then flow through the rest of the type system. But... that's a lot of work.

Understood.  Feels very marginal in value.  Maybe one line in 10k i would find this sort of value.

**Reply by FaustVX** [reply]
@RikkiGibson 

Similar to your idea:
I really want to be able to call instance methods (end extensions ?) in switch cases
```csharp
var s = "Hello C#";
switch (s)
{
  case .StartWith("Hello"):
    // Do something
    break;
  case .EndWith("C#"):
    // Do something else
    break;
}
```

**Reply by bernd5** [reply]
@FaustVX

Today you can write:

```csharp
var s = "Hello C#";
switch (s)
{
  case var x when x.StartsWith("Hello"):
    // Do something
    break;
  case var x when x.EndsWith("C#"):
    // Do something else
    break;
}
```

But I understand you want to omit the var pattern...

**Reply by FaustVX** [reply]
Yes, I can write like that, but I find the `case var x when x ...` kind of an antipattern, really not so easy to read, and not very esthetic.

**Reply by RikkiGibson** [reply]
I expected `case _ when ...` to be possible. I was surprised at what I actually got: [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAcoKgW1QOzgBsiBjACwlIGsBLfAcwB8A3LAAmXYF52AiABIQSAe3YBhAMR8A3AFgAUMgDutGBXYAKZAEpFAb0Xt2pKMgjsA+u2WV8nAHQBlGFhjIA6mvKbBwoiJ8OiBGxuwA9OHsACJiyCI4EDDk9AyhxsBIUNTyCsam5lY2do4AovgAJp7evlJBIXlhkTFxCUkpjOzC5unsmdA5igC+QA=)

**Reply by FaustVX** [reply]
Oh, ok, I'm surprised too.
But even it's that work (and it's similar with the `var x when x` pattern), IMO, it's kinda worse, in the sense of: I telling the compiler I don't care about that actual value, but I care about it in the `when` clause (and I have to rewrite the `s`)

**Reply by sab39** [reply]
How about `case when _.StartsWith(...)`. Obviously not legal today, but the `when` would give a handy hint that what follows is a boolean expression rather than a pattern, and it's not normally legal to dereference a discard so I don't think it would be ambiguous (unless there's actually something called `_` in scope, in which case all discards are potentially affected anyway)

**Reply by bernd5** [reply]
Yes, a discard pattern in switch statements would be nice.
Why we can only use a var pattern with a discard designation `var _` or `default`?
I assume it is done this way for backward compatibility reasons (_ could bind to something).

**Reply by FaustVX** [reply]
@sab39 
As your proposal and mine, are illegal today, I like much my version :)
It's a simpler and more intuitive syntax.
But I'm okay with your syntax as well.

**Reply by HaloFour** [reply]
I'd rather [active patterns](https://github.com/dotnet/csharplang/issues/277).  Trying to cram general method calls or other expressions into the middle of patterns or switch cases just doesn't feel right to me.

**Reply by 333fred** [reply]
I'm personally of the opinion that we have list patterns, dictionary patterns, and active patterns left to do. Once those 3 are added, I think we have all the common cases simple, and anything else can use an active pattern.

**Reply by sab39** [reply]
The linked issue about active patterns was closed in 2017 and last touched in 2019... is there something more recent about the current state of LDM thinking about active patterns?

**Reply by HaloFour** [reply]
Here's a more recent proposal: https://github.com/dotnet/csharplang/issues/4131

**Reply by 333fred** [reply]
I just realized that one didn't have a champion. I've assigned myself.

**Reply by FaustVX** [reply]
Yes, I forgot about the Active pattern, I'm really looking for that, way better than my proposal 😄
